lasagne/layers/recurrent.py

```diff
@@ -213,12 +230,47 @@ def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
             raise ValueError("Input sequence length cannot be specified as "
                              "None when unroll_scan is True")
 
-        # Check that output shapes match
-        if input_to_hidden.output_shape != hidden_to_hidden.output_shape:
+        # Check that the input_to_hidden connection can appropriately handle
+        # a first dimension of input_shape[0]*input_shape[1] when we will
+        # precompute the input dot product
+        if (self.precompute_input and
+                input_to_hidden.output_shape[0] is not None and
+                (input_to_hidden.output_shape[0] !=
+                 input_shape[0]*input_shape[1])):
```

Comment:
You also need to add `and input_shape[0] is not None and input_shape[1] is not None`, otherwise you can't do the multiplication and have to skip the early check.